openeogeotrellis/user_defined_process_repository.py

```diff
@@ -82,16 +84,32 @@ def delete(self, user_id: str, process_id: str) -> None:
 
     @contextlib.contextmanager
     def _zk_client(self):
-        kz_retry = KazooRetry(max_tries=10, delay=0.5, backoff=2)
-        zk = KazooClient(hosts=self._hosts,connection_retry=kz_retry,
-                 command_retry=kz_retry, timeout=3.0)
-        zk.start(timeout=15.0)
+        if not self._zk_client_reuse or self._zk_client_cache is None:
```

Comment:
Since `self._zk_client_cache is None` is part of the new initialization I would limit the check tot that as it is more readable imho
```suggestion
        if not self._zk_client_cache is None:
```

Reply:
If it is to protect against programming errors I would not do it in the same if branch but rather split it out and assert:
```
if not self._zk_client_reuse:
  assert self._zk_client_cache is None, "cache is only for client reuse so must be None if re-use is disabled"
```

Reply:
I indeed considered initially what you suggested there, but I don't completely agree about the readability. 
It's true that combining `not` and `or` in a condition can be a bit tricky, but I kind of like what the current condition clearly expresses: create a new client in these two situations:
- reuse is disabled
- reuse is enabled, but it's the first time we're here

By removing the mention of `zk_client_reuse` in the condition, you obfuscate how that reasoning and you have to understand how `_zk_client_cache` is handled to fully grasp that.

Would it be better to flip the argument `_zk_client_reuse` to e.g. `_zk_recreate_client` so that the condition becomes (without a `not`):

```
if  self._zk_recreate_client or self._zk_client_cache is None:
```

another trick is introducing a new variable to clarify the condition:
```
need_to_create_new_client = not self._zk_client_reuse or self._zk_client_cache is None
if need_to_create_new_client:
```

Reply:
Either suggestion is an improvement and both would be OK to me. 
I might lean slightly to the 2nd but I cannot express clearly why so either would work for me